### PR TITLE
Use dynamic client to consolidate TaskRun and PipelineRun logic.

### LINKED ISSUE #102: Look into using dynamic clients for type agnostic watching
We have a lot of duplication in the watcher code, mostly because we don't have a generic type to use for TaskRuns and PipelineRuns. 

[Dynamic clients](https://pkg.go.dev/k8s.io/client-go/dynamic) might be a good compromise here to use [Unstructured](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#Unstructured) types to support anything that is Object-like.

### PR DESCRIPTION
**Pull request by wlynch**
Most of the logic between TaskRun and PipelineRuns was near identical,
but since we were using typed clients there was no easy way to handle
these with a single library. This PR changes the implementation to use a
dynamic client, changing the controllers to use the same code that
target different resource versions.

The goal is to use this to make it easier to add additional types (e.g.
v1alpha1) without massive duplication of logic for each type.

Fixes #102

### CODE REVIEW

File: pkg/watcher/reconciler/dynamic/dynamic.go [hunk 1]
```diff
@@ -12,52 +12,60 @@
...
 	log.With(zap.String("key", key))
 
 	namespace, name, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
 		log.Errorf("invalid resource key: %s", key)
 		return nil
 	}
-
+	log.Infof("key: %s, namespace: %s, name: %s, gvr: %v", key, namespace, name, r.gvr)
 	// Lookup current TaskRun.
```
Comment (ImJasonH) [inline pkg/watcher/reconciler/dynamic/dynamic.go:56]: This comment is out of date now.

File: pkg/watcher/reconciler/reconciler_test.go [hunk 2]
```diff
@@ -57,102 +62,108 @@ func TestController(t *testing.T) {
 
 	pipeline := fakepipelineclient.Get(ctx)
 	t.Run("taskrun", func(t *testing.T) {
-		reconcileTaskRun(ctx, t, pipeline)
-	})
-	t.Run("pipelinerun", func(t *testing.T) {
-		reconcilePipelineRun(ctx, t, pipeline)
-	})
+		//reconcileTaskRun(ctx, t, pipeline)
```
Comment (ImJasonH) [inline pkg/watcher/reconciler/reconciler_test.go:65]: I think this was unintentional.

**Review by ImJasonH — COMMENTED** [review]
Say, this should make it easier to watch `Run`s too 😄

**Comment by sbwsg** [comment]
/lgtm

I'm adding a hold to give you a chance to review the other comments prior to merge. 

/hold

File: pkg/watcher/reconciler/dynamic/dynamic.go (hunk 1 above)
Comment (wlynch) [reply pkg/watcher/reconciler/dynamic/dynamic.go]: Changed references from TaskRun -> Object.

File: pkg/watcher/reconciler/reconciler_test.go (hunk 2 above)
Comment (wlynch) [reply pkg/watcher/reconciler/reconciler_test.go]: Removed.

**Comment by wlynch** [comment]
> Say, this should make it easier to watch `Run`s too 😄

You've seen through my plan 😎

**Pull request merged**